File: app_edools/app/models/school.rb
class School < ActiveRecord::Base
  has_many :courses
  has_many :students

  validates_presence_of :name
  validates_presence_of :email_responsible
  validates_presence_of :pitch
  validates_uniqueness_of :name
  validates_uniqueness_of :email_responsible
  validates_uniqueness_of :subdomain
  validates_format_of :subdomain, with: /[A-z\d]*.[A-z\d]*\.edools.com/
  validates_format_of :email_responsible, with: /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\z/i, on: :create
end

File: backend/src/app/models/School.js
import Sequelize, { Model } from 'sequelize';
import bcrypt from 'bcryptjs';

class School extends Model {
  static init(sequelize) {
    super.init(
      {
        name: Sequelize.STRING,
        email: Sequelize.STRING,
        password: Sequelize.VIRTUAL,
        password_hash: Sequelize.STRING,
        pitch: Sequelize.STRING,
        subdomain: Sequelize.STRING,
        active: Sequelize.BOOLEAN,
      },
      {
        sequelize,
      }
    );

    this.addHook('beforeSave', async school => {
      if (school.password) {
        school.password_hash = await bcrypt.hash(school.password, 8);
      }
    });

    return this;
  }

  static associate(models) {
    this.hasMany(models.Course, { foreignKey: 'school_id', as: 'courses' });
    this.hasMany(models.Student, { foreignKey: 'school_id', as: 'students' });
  }

  checkPassword(password) {
    return bcrypt.compare(password, this.password_hash);
  }

  static checkSubdomain(subdomain) {
    return /^[a-zA-Z0-9-_]+$/.test(subdomain);
  }
}

export default School;
